xapian-letor/ranker/ranker.cc

```diff
@@ -22,13 +22,280 @@
+	fv.set_label(label);
+	train_file.ignore(); // <label>* *qid:<xxx> n:<fval> #docid=<xxx>
+	string qid;
+	train_file >> qid;   // <label> *qid:<xxx>* n:<fval> #docid=<xxx>
+	qid = qid.substr(qid.find(":") + 1, qid.length()); // <label> qid:*<xxx>* n:<fval> #docid=<xxx>
+	train_file.ignore(); // <label> qid:<xxx>* *n:<fval> #docid=<xxx>
+
+	std::vector<double> fvals;
+	while (train_file.peek() != '#') { // read till '#docid' is found
+	    int feature_index;
+	    double feature_value;
+	    train_file >> feature_index; // <label> qid:<xxx> *n*:<fval> #docid=<xxx>
+	    train_file.ignore();         // <label> qid:<xxx> n*:*<fval> #docid=<xxx>
+	    train_file >> feature_value; // <label> qid:<xxx> n:*<fval>* #docid=<xxx>
+	    fvals.push_back(feature_value);
+	    train_file.ignore();         // <label> qid:<xxx> n:<fval>* *#docid=<xxx>
+	}
+	fv.set_fvals(fvals);
+
+	string did;
+	train_file >> did;               // <label> qid:<xxx> n:<fval> *#docid=<xxx>*
+	did = did.substr(did.find("=") + 1, did.length()); // #docid=*<xxx>*
+	Xapian::docid docid = (Xapian::docid) atoi(did.c_str());
+	fv.set_did(docid);
+	train_file.ignore();
+
+	fv.set_score(0);
+	fvv.push_back(fv);
+    }
+    return fvv;
+}
+
+static int
+getlabel(const Document & doc, const std::string & qid)
+{
+    int label = -1;
+    string id = std::to_string(doc.get_docid());
+    map<string, map<string, int> >::const_iterator outerit;
+    map<string, int>::const_iterator innerit;
+
+    outerit = qrel.find(qid);
+    if (outerit != qrel.end()) {
+	innerit = outerit->second.find(id);
+	if (innerit != outerit->second.end()) {
+	    label = innerit->second;
+	}
+    }
+    return label;
+}
+
+static map<string, map<string, int>>
+load_relevance(const std::string & qrel_file)
+{
+    map<string, map<string, int>> qrel1;     // < qid, <docid, relevance_judgement> >
+
+    string inLine;
+    ifstream myfile (qrel_file.c_str(), ifstream::in);
+    if (!myfile.good()) {
+	throw Xapian::FileNotFoundError("No Qrel file found. Check path.");
+    }
+    int qrel_count = 0;
+    if (myfile.is_open()) {
+	while (myfile.good()) {
+	    getline(myfile, inLine);        // read a file line by line
+	    if (inLine.empty()) {
+		break;
+	    }
+	    ++qrel_count;
+	    vector<string> token;
+	    size_t j = 0;
+	    while (j < inLine.size()) {
+		size_t i = inLine.find_first_not_of(' ', j);
+		if (i == string::npos) break;
+		j = inLine.find_first_of(' ', i);
+		token.push_back(inLine.substr(i, j - i));
+	    }
+	    // Exceptions for parse errors
+	    if (token.size() != 4 || token[1] != "Q0") {
+		throw LetorParseError("Could not parse Qrel file at line:" + Xapian::Internal::str(qrel_count));
+	    }
+	    // Exception if relevance label is not a number
+	    int label;
+	    char * end;
+	    label = int(strtol(token[3].c_str(), &end, 10));
+	    if (*end) {
+		throw LetorParseError("Could not parse relevance label in Qrel file at line:"
+				       + Xapian::Internal::str(qrel_count));
```

Comment:
Just use `str()` (the `Xapian::Internal::` qualification isn't needed, and is just an implementation detail really, so could change in the future).